src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -198,22 +205,25 @@ def _osd_flags():
     def list(self):
         return self._osd_flags()
 
-    def bulk_set(self, flags):
-        """
-        The `recovery_deletes`, `sortbitwise` and `pglog_hardlimit` flags cannot be unset.
-        `purged_snapshots` cannot even be set. It is therefore required to at
-        least include those four flags for a successful operation.
-        """
+    def bulk_delete(self, flags):
```

Comment:
Using `PUT/DELETE` to mimic `set/unset` directives is an interesting approach, but not very REST-y. `DELETE /osd/flags` should involve deleting the flag resource collection.

Reply:
well, unless you want to have different endpoints for `set` and `unset`, I'm not sure how you'd get away with anything else besides doing partial `PUT` and `DELETE`. Alternatively, have the caller providing more context so we can decide whether we're doing a partial `PUT` for `set` or `unset`.

As for the semantics not being very REST-y, at least for `PUT` this is what I found in [1].

> If the Request-URI refers to an already existing resource – an update operation will happen, otherwise create operation should happen [...]

Which I'm interpreting an `update` operation as being something less than a create, but the actual definition seems a bit up in the air. To be perfectly honest, my only experience with REST api's has been the Hue's, and they do use `PUT` for partial updates of state on an endpoint.

[1] https://restfulapi.net/rest-put-vs-post/

Reply:
The way it's defined here, both `PUT` and `DELETE` are performing partial updates:
- `PUT` as per HTTP 1.1 should not perform partial updates. `PATCH` should be used instead,
- `DELETE` technically should involve doing some kind of resource deallocation, rather than partially updating resource metadata.

Part of the issue might come from the fact that we perhaps have given flags the entity of a resource when it should be rather just resource metadata of the OSD resource. Alternatively, we could convert them to a collection endpoint, being every flag a named sub-document of the `OSDFlagsCollection` (`/osd/flags/<flag_name>`), but that's probably an overkill.